### DISCUSSION TITLE
is the presentation layer should be an abstraction per framework?

### DISCUSSION THREAD

**Discussion by kattsushi** [guiseek/nx-clean#61] [Q&A]
It is possible to generate an abstraction that is for each supported framework and thus improve performance. For example, Angular recently introduced Signals for the reactivity of values in the UI, so in this case it should return signals values but also in the case of using vanilla web framework use Promises instead of Observables, this layer could be added in the mappers and that the types are automatically inferred

**Comment by kattsushi** [comment]
pointing out this PR https://github.com/guiseek/nx-clean/pull/54 implementation for React

**Comment by guiseek** [comment]
> It is possible to generate an abstraction that is for each supported framework and thus improve performance. For example, Angular recently introduced Signals for the reactivity of values in the UI, so in this case it should return signals values but also in the case of using vanilla web framework use Promises instead of Observables, this layer could be added in the mappers and that the types are automatically inferred

improve performance? I don't understand, can you be more specific @kattsushi ?

and what is the reason for replacing observables with signals?

**Reply by kattsushi** [reply]
yes my idea is to have something like the PR #54 for react but for Angular and also have the package for others frameworks as well keeping the developer experience as to near as possible per framework at the moment to get and implement the Presenter layer, in the case of Angular leaving this post [Angular 16 Reactivity Performance](https://www.itmagination.com/blog/web-dev-angular-16-reactivity-performance-signals) Angular 16 introduce Signals and move Stand Alone Api to ready to use in production is moving to have reactivity in more grained way 

About Rxjs i have two concerns, first keeping the DX my idea replacing with promises ( or can be optional by parameter ) ideally all Clean Architecture should be framework agnostic and Rxjs is a too angular way to work in terms of DX son the Domains, Presentation, and Data layers probably the best is moving near to the language and only use Javascript Apis ( compatible with browsers and runtimes ) and use ie. Promises.

Rxjs should be optional and parametrized at the moment of generate the libraries and we can have others tools and libraries we can do support for example Zod that is a powerful tool to create schemas, do contracts and validate in runtime for mappers for javascript and typescript but also should be optional, the idea is not opinionate in tools only opinionate applying the separation of concerns and the best practices of Clean Architecture 

But in another hand in the Port per framework specific beside the Presentation layer use the tools that framework support ie.

- Angular can use just Observables or Signals or Ngrx Store Component with Signals/Observables
- React can use Hooks with useState or useReducer even Redux
- Qwik can use Hooks with Signals
- Svelte can use Writable Stores
- Web Framework, Api framework or even Other SubDomain can use by the default presenter based in Promises or Observables 

each framework have utils to get the presenter from @nx-clean/react, @nx-clean/angular, @nx-clean/... etc 

I know we have a bit away to get there but the final point is for this nx plugin to become widely used, it needs to be framework agnostic, not only in the implementation but also in terms of DX, but optionally supports what we already have at the moment, for example Rxjs and observables but by default on the side of Clean Architecture be with Promises,

supporting the most popular frameworks and maintains the development experience that each framework already has and uses tools to connect that DX with the Presenter in Clean Arch.

In this way we can promote the use and good practices of clean architecture, separation of concerns, sharing business rules between different backend or frontend frameworks.
